### PR TITLE
Throw if rewind to last checkpoint fails in AbstractKafkaBasedConnectorTask

### PR DESCRIPTION
**Pull request by somandal**
The AbstractKafkaBasedConnectorTask attempts to rewind TopicPartitions to the last checkpoint if any send failures are seen. To avoid data loss, we should thrown an exception if rewind fails, since we may accidentally commit newer offsets for data we were not able to send successfully (e.g. calling KafkaConsumer pause() and then calling commitSync() will commit offsets for paused TopicPartitions).

Important: DO NOT REPORT SECURITY ISSUES DIRECTLY ON GITHUB.  
For reporting security issues and contributing security fixes,  
please, email security@linkedin.com instead, as described in  
the contribution guidelines.

Please, take a minute to review the contribution guidelines at:  
https://github.com/linkedin/Brooklin/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: datastream-kafka-connector/src/test/java/com/linkedin/datastream/connectors/kafka/mirrormaker/TestKafkaMirrorMakerConnectorTask.java [hunk 1]
```diff
@@ -344,12 +345,24 @@ protected void postShutdownHook() {
...
 
+    @Override
+    protected void seekToLastCheckpoint(Set<TopicPartition> topicPartitions) {
+      if (_failOnSeekToLastCheckpoint) {
+        throw new KafkaException("KafkaException: failed to seek");
+      }
+      super.seekToLastCheckpoint(topicPartitions);
+    }
+
+    void setFailOnSeekToLastCheckpoint() {
```
Comment (DEEPTHIKORAT) [inline datastream-kafka-connector/src/test/java/com/linkedin/datastream/connectors/kafka/mirrormaker/TestKafkaMirrorMakerConnectorTask.java:360]: nit: wondering if the setter can take true/false to allow for stop failing send failures also.

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java [hunk 2]
```diff
@@ -282,15 +284,21 @@ protected void rewindAndPausePartitionOnException(TopicPartition srcTopicPartiti
...
 
   protected void rewindAndPausePartitionsOnSendException() {
-    // For all topic partitions which have seen send exceptions, attempt to rewind them to the last checkpoint
+    // For all topic partitions which have seen send exceptions, attempt to rewind them to the last checkpoint.
+    // The outcome of the rewind can fall into three categories:
+    // 1) The rewind is successful and the Exception returned by the SendCallback is transient. This TopicPartition is
+    //    not added to the auto-pause list.
+    // 2) The rewind is successful and the Exception returned by the SendCallback is non-transient. This TopicPartition
+    //    is added to the auto-pause list with SEND_ERROR as the reason.
+    // 3) The rewind itself failed. An exception is thrown, and the connector task is brought down to avoid wrongfully
```
Comment (vmaheshw) [inline datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:293]: nit: do you need both ',' and 'and'

Comment (vmaheshw) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:293]: also, can you rephrase this comment may be: the connector task is brought down to avoid committing wrong checkpoints.

File: datastream-kafka-connector/src/test/java/com/linkedin/datastream/connectors/kafka/mirrormaker/KafkaMirrorMakerConnectorTestUtils.java [hunk 3]
```diff
@@ -132,9 +132,15 @@ static KafkaMirrorMakerConnectorTask createFlushlessKafkaMirrorMakerConnectorTas
 
   static Thread runKafkaMirrorMakerConnectorTask(KafkaMirrorMakerConnectorTask connectorTask)
       throws InterruptedException {
+    return runKafkaMirrorMakerConnectorTask(connectorTask, (t, e) -> Assert.fail("connector thread died", e));
+  }
+
+  static Thread runKafkaMirrorMakerConnectorTask(KafkaMirrorMakerConnectorTask connectorTask,
+      Thread.UncaughtExceptionHandler exceptionHandler)
+      throws InterruptedException {
```
Comment (vmaheshw) [inline datastream-kafka-connector/src/test/java/com/linkedin/datastream/connectors/kafka/mirrormaker/KafkaMirrorMakerConnectorTestUtils.java:140]: nit, can you merge the two lines

File: datastream-kafka-connector/src/test/java/com/linkedin/datastream/connectors/kafka/mirrormaker/TestKafkaMirrorMakerConnectorTask.java [hunk 4]
```diff
@@ -821,6 +834,56 @@ public void testAutoPauseAndResumeOnSendFailure() throws Exception {
...
+    // create a datastream to consume from topics ending in "Pizza"
+    Datastream datastream = KafkaMirrorMakerConnectorTestUtils.createDatastream("pizzaStream", _broker, "\\w+Pizza");
+
+    DatastreamTaskImpl task = new DatastreamTaskImpl(Collections.singletonList(datastream));
+    // create event producer that fails on 3rd event (of 5)
+    MockDatastreamEventProducer datastreamProducer =
+        new MockDatastreamEventProducer((r) -> new String((byte[]) r.getEvents().get(0).key().get()).equals("key-2"));
+    task.setEventProducer(datastreamProducer);
+
+    ZkAdapter zkAdatper = new ZkAdapter(_kafkaCluster.getZkConnection(), "testCluster", null,
```
Comment (vmaheshw) [inline datastream-kafka-connector/src/test/java/com/linkedin/datastream/connectors/kafka/mirrormaker/TestKafkaMirrorMakerConnectorTask.java:851]: typo zkAdatper

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java (hunk 2 above)
Comment (somandal) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:293]: done

File: datastream-kafka-connector/src/test/java/com/linkedin/datastream/connectors/kafka/mirrormaker/KafkaMirrorMakerConnectorTestUtils.java (hunk 3 above)
Comment (somandal) [reply datastream-kafka-connector/src/test/java/com/linkedin/datastream/connectors/kafka/mirrormaker/KafkaMirrorMakerConnectorTestUtils.java:140]: done

File: datastream-kafka-connector/src/test/java/com/linkedin/datastream/connectors/kafka/mirrormaker/TestKafkaMirrorMakerConnectorTask.java (hunk 4 above)
Comment (somandal) [reply datastream-kafka-connector/src/test/java/com/linkedin/datastream/connectors/kafka/mirrormaker/TestKafkaMirrorMakerConnectorTask.java:851]: done

File: datastream-kafka-connector/src/test/java/com/linkedin/datastream/connectors/kafka/mirrormaker/TestKafkaMirrorMakerConnectorTask.java (hunk 1 above)
Comment (somandal) [reply datastream-kafka-connector/src/test/java/com/linkedin/datastream/connectors/kafka/mirrormaker/TestKafkaMirrorMakerConnectorTask.java]: done
